File: indexes/cpp/src/npy.cpp
#include "npy.hpp"

#include <cstring>
#include <fstream>
#include <stdexcept>

namespace vro::npy {
namespace {

struct Header {
    std::string descr;
    std::size_t rows = 0;
    std::size_t cols = 0;
};

[[noreturn]] void fail(const std::string& path, const std::string& msg) {
    throw std::runtime_error(path + ": " + msg);
}

// Returns the text after "'key':" in the header dict, with leading spaces removed.
std::string value_after(const std::string& dict, const std::string& key, const std::string& path) {
    std::string pattern = "'" + key + "':";
    auto pos = dict.find(pattern);
    if (pos == std::string::npos) fail(path, "header has no '" + key + "'");
    pos += pattern.size();
    while (pos < dict.size() && dict[pos] == ' ') ++pos;
    return dict.substr(pos);
}

Header parse_header(const std::string& dict, const std::string& path) {
    Header h;
    std::string d = value_after(dict, "descr", path);
    if (d.empty() || d[0] != '\'') fail(path, "bad 'descr'");
    h.descr = d.substr(1, d.find('\'', 1) - 1);

    std::string f = value_after(dict, "fortran_order", path);
    if (f.rfind("False", 0) != 0) fail(path, "fortran_order must be False");

    std::string s = value_after(dict, "shape", path);
    if (s.empty() || s[0] != '(') fail(path, "bad 'shape'");
    std::string inner = s.substr(1, s.find(')') - 1);
    std::vector<std::size_t> dims;
    std::size_t i = 0;
    while (i < inner.size()) {
        while (i < inner.size() && (inner[i] == ' ' || inner[i] == ',')) ++i;
        if (i >= inner.size()) break;
        std::size_t used = 0;
        dims.push_back(std::stoull(inner.substr(i), &used));
        i += used;
    }
    if (dims.size() != 2) fail(path, "expected a 2-D array");
    h.rows = dims[0];
    h.cols = dims[1];
    return h;
}

// Opens the file, checks magic and version, parses the header. Leaves the
// stream at the data offset (10 + HLEN).
Header open_and_parse(std::ifstream& in, const std::string& path) {
    in.open(path, std::ios::binary);
    if (!in) fail(path, "cannot open file");
    unsigned char pre[10];
    if (!in.read(reinterpret_cast<char*>(pre), 10)) fail(path, "file too short");
    if (std::memcmp(pre, "\x93NUMPY", 6) != 0) fail(path, "not a .npy file (bad magic)");
    if (pre[6] != 1 || pre[7] != 0) fail(path, "only .npy version 1.0 is supported");
    std::size_t hlen = static_cast<std::size_t>(pre[8]) | (static_cast<std::size_t>(pre[9]) << 8);
    std::string dict(hlen, '\0');
    if (!in.read(dict.data(), static_cast<std::streamsize>(hlen))) fail(path, "header truncated");
    return parse_header(dict, path);
}

template <typename T>
void read_data(std::ifstream& in, std::vector<T>& out, std::size_t count, const std::string& path) {
    out.resize(count);
    auto bytes = static_cast<std::streamsize>(count * sizeof(T));
    if (!in.read(reinterpret_cast<char*>(out.data()), bytes)) fail(path, "data truncated");
}

}  // namespace

Matrix read_f32(const std::string& path, std::size_t max_rows) {
    std::ifstream in;
    Header h = open_and_parse(in, path);
    if (h.descr != "<f4") fail(path, "descr is '" + h.descr + "', expected '<f4'");
    Matrix m;
    m.rows = (max_rows > 0 && max_rows < h.rows) ? max_rows : h.rows;
    m.dim = h.cols;
    read_data(in, m.data, m.rows * m.dim, path);
    return m;
}

Int64Array read_i64(const std::string& path) {
    std::ifstream in;
    Header h = open_and_parse(in, path);
    if (h.descr != "<i8") fail(path, "descr is '" + h.descr + "', expected '<i8'");
    Int64Array a;
    a.rows = h.rows;
    a.cols = h.cols;
    read_data(in, a.data, a.rows * a.cols, path);
    return a;
}

}  // namespace vro::npy

File: indexes/rust/src/npy.rs
//! A hand-written reader for NumPy `.npy` files, format version 1.0 (CONTRACT 1.1).

use std::fs::File;
use std::io::{BufReader, Read, Seek, SeekFrom};
use std::path::Path;

const MAGIC: &[u8; 6] = b"\x93NUMPY";
/// Bytes converted per read, so a large file never needs a second full-size buffer.
const CHUNK_BYTES: usize = 1 << 22;

/// A row-major 2-D array in one contiguous buffer.
#[derive(Debug, Clone, PartialEq)]
pub struct Array2<T> {
    pub data: Vec<T>,
    pub rows: usize,
    pub cols: usize,
}

/// A float32 matrix: the corpus or the queries.
pub type Matrix = Array2<f32>;

impl<T> Array2<T> {
    #[inline]
    pub fn row(&self, i: usize) -> &[T] {
        &self.data[i * self.cols..(i + 1) * self.cols]
    }
}

/// The parsed header fields that the reader checks.
#[derive(Debug, Clone, PartialEq)]
pub struct Header {
    pub descr: String,
    pub fortran_order: bool,
    pub shape: Vec<usize>,
    /// Offset of the first data byte: 10 + HLEN.
    pub data_offset: u64,
}

/// Reads a `<f4` 2-D array. `max_rows` limits the rows read (`--limit`).
pub fn read_f32(path: &Path, max_rows: Option<usize>) -> Result<Matrix, String> {
    read_array(path, "<f4", max_rows, |b: &[u8; 4]| {
        f32::from_le_bytes([b[0], b[1], b[2], b[3]])
    })
}

/// Reads a `<i8` (int64) 2-D array, for example `ground_truth.npy`.
pub fn read_i64(path: &Path, max_rows: Option<usize>) -> Result<Array2<i64>, String> {
    read_array(path, "<i8", max_rows, |b: &[u8; 8]| {
        i64::from_le_bytes([b[0], b[1], b[2], b[3], b[4], b[5], b[6], b[7]])
    })
}

fn read_array<T, const W: usize>(
    path: &Path,
    descr: &str,
    max_rows: Option<usize>,
    convert: impl Fn(&[u8; W]) -> T,
) -> Result<Array2<T>, String> {
    let err = |msg: String| format!("{}: {msg}", path.display());
    let file = File::open(path).map_err(|e| err(e.to_string()))?;
    let mut reader = BufReader::new(file);
    let header = read_header(&mut reader).map_err(err)?;
    if header.descr != descr {
        return Err(err(format!(
            "descr is '{}', expected '{descr}'",
            header.descr
        )));
    }
    let (total_rows, cols) = match header.shape[..] {
        [r, c] => (r, c),
        _ => return Err(err(format!("shape {:?} is not 2-D", header.shape))),
    };
    let rows = max_rows.map_or(total_rows, |m| m.min(total_rows));
    reader
        .seek(SeekFrom::Start(header.data_offset))
        .map_err(|e| err(e.to_string()))?;
    let data = read_values(&mut reader, rows * cols, convert).map_err(err)?;
    Ok(Array2 { data, rows, cols })
}

/// Reads `count` little-endian values of `W` bytes each, in chunks.
fn read_values<T, const W: usize>(
    reader: &mut impl Read,
    count: usize,
    convert: impl Fn(&[u8; W]) -> T,
) -> Result<Vec<T>, String> {
    let mut out = Vec::with_capacity(count);
    let mut buf = vec![0u8; CHUNK_BYTES - CHUNK_BYTES % W];
    let mut left = count * W;
    while left > 0 {
        let n = left.min(buf.len());
        reader
            .read_exact(&mut buf[..n])
            .map_err(|e| format!("data is shorter than the shape says: {e}"))?;
        out.extend(buf[..n].chunks_exact(W).map(|c| {
            let bytes: &[u8; W] = c.try_into().expect("chunk has W bytes");
            convert(bytes)
        }));
        left -= n;
    }
    Ok(out)
}

/// Reads and parses the magic string, version, and header dict.
pub fn read_header(reader: &mut impl Read) -> Result<Header, String> {
    let mut pre = [0u8; 10];
    reader
        .read_exact(&mut pre)
        .map_err(|e| format!("cannot read header: {e}"))?;
    if &pre[..6] != MAGIC {
        return Err("not a .npy file (bad magic string)".into());
    }
    if pre[6] != 1 {
        return Err(format!(
            "npy version {}.{} is not supported, need 1.0",
            pre[6], pre[7]
        ));
    }
    let hlen = u16::from_le_bytes([pre[8], pre[9]]) as usize;
    let mut dict = vec![0u8; hlen];
    reader
        .read_exact(&mut dict)
        .map_err(|e| format!("cannot read header dict: {e}"))?;
    let dict = String::from_utf8(dict).map_err(|_| "header dict is not ASCII".to_string())?;
    parse_dict(&dict, 10 + hlen as u64)
}

/// Parses `{'descr': '<f4', 'fortran_order': False, 'shape': (N, D), }`.
fn parse_dict(dict: &str, data_offset: u64) -> Result<Header, String> {
    let descr = value_after(dict, "descr")?;
    let descr = descr
        .strip_prefix('\'')
        .and_then(|s| s.split('\'').next())
        .ok_or_else(|| format!("bad descr in header: {dict}"))?
        .to_string();
    let fortran = value_after(dict, "fortran_order")?;
    let fortran_order = if fortran.starts_with("False") {
        false
    } else if fortran.starts_with("True") {
        return Err("fortran_order is True; only C order is supported".into());
    } else {
        return Err(format!("bad fortran_order in header: {dict}"));
    };
    let shape = value_after(dict, "shape")?;
    let inner = shape
        .strip_prefix('(')
        .and_then(|s| s.split(')').next())
        .ok_or_else(|| format!("bad shape in header: {dict}"))?;
    let shape = inner
        .split(',')
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .map(|s| {
            s.parse::<usize>()
                .map_err(|_| format!("bad shape entry '{s}'"))
        })
        .collect::<Result<Vec<_>, _>>()?;
    Ok(Header {
        descr,
        fortran_order,
        shape,
        data_offset,
    })
}

/// Returns the text after `'key':`, with leading spaces removed.
fn value_after<'a>(dict: &'a str, key: &str) -> Result<&'a str, String> {
    let pat = format!("'{key}':");
    let start = dict
        .find(&pat)
        .ok_or_else(|| format!("header has no '{key}': {dict}"))?;
    Ok(dict[start + pat.len()..].trim_start())
}
